lib/services/local/generic/generic_wrapper_test.go

```diff
@@ -0,0 +1,225 @@
+	}
+	cfg, err := services.CollectOptions(opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var r testResource153
+	if err := utils.FastUnmarshal(data, &r); err != nil {
+		return nil, trace.BadParameter(err.Error())
+	}
+
+	if r.Metadata == nil {
+		r.Metadata = &headerv1.Metadata{}
+	}
+
+	if cfg.ID != 0 {
+		//nolint:staticcheck // SA1019. Deprecated, but still needed.
+		r.Metadata.Id = cfg.ID
+	}
+	if cfg.Revision != "" {
+		r.Metadata.Revision = cfg.Revision
+	}
+	if !cfg.Expires.IsZero() {
+		r.Metadata.Expires = timestamppb.New(cfg.Expires)
+	}
+	return &r, nil
+}
+
+// TestGenericWrapperCRUD tests backend operations with the generic service.
+func TestGenericWrapperCRUD(t *testing.T) {
+	ctx := context.Background()
+
+	memBackend, err := memory.New(memory.Config{
+		Context: ctx,
+		Clock:   clockwork.NewFakeClock(),
+	})
+	require.NoError(t, err)
+
+	const backendPrefix = "generic_prefix"
+
+	service, err := NewServiceWrapper[*testResource153](memBackend,
+		"generic resource",
+		backendPrefix,
+		marshalResource153,
+		unmarshalResource153)
+	require.NoError(t, err)
+
+	// Create a couple test resources.
+	r1 := newTestResource153("r1")
+	r2 := newTestResource153("r2")
+
+	// sanity check: marshal/unmarshal round trips
+	r1data, err := marshalResource153(r1)
+	require.NoError(t, err)
+	r1clone, err := unmarshalResource153(r1data)
+	require.NoError(t, err)
+	require.Equal(t, r1, r1clone)
+
+	// Initially we expect no resources.
+	out, nextToken, err := service.ListResources(ctx, 200, "")
+	require.NoError(t, err)
+	require.Empty(t, nextToken)
+	require.Empty(t, out)
+
+	// Create both resources.
+	r1, err = service.CreateResource(ctx, r1)
+	require.NoError(t, err)
+	r2, err = service.CreateResource(ctx, r2)
+	require.NoError(t, err)
+
+	require.NotEmpty(t, r1.GetMetadata().GetRevision())
+	require.NotEmpty(t, r2.GetMetadata().GetRevision())
+	require.NotEqual(t, r1.GetMetadata().GetRevision(), r2.GetMetadata().GetRevision())
+
+	// Fetch all resources using paging default.
+	out, nextToken, err = service.ListResources(ctx, 0, "")
+	require.NoError(t, err)
+	require.Empty(t, nextToken)
+	require.NotEmpty(t, out)
+	require.Equal(t, []*testResource153{r1, r2}, out)
+
+	// Fetch a paginated list of resources.
+	paginatedOut := make([]*testResource153, 0, 2)
+	numPages := 0
+	for {
+		numPages++
+		out, nextToken, err = service.ListResources(ctx, 1, nextToken)
+		require.NoError(t, err)
+
+		paginatedOut = append(paginatedOut, out...)
+		if nextToken == "" {
+			break
+		}
+	}
+
+	require.Equal(t, 2, numPages)
+	require.Equal(t, []*testResource153{r1, r2}, paginatedOut)
+
+	// Fetch a specific service provider.
+	r, err := service.GetResource(ctx, r2.GetMetadata().GetName())
+	require.NoError(t, err)
+	require.Equal(t, r2, r)
+
+	// Try to fetch a resource that doesn't exist.
+	_, err = service.GetResource(ctx, "doesnotexist")
+	require.ErrorIs(t, err, trace.NotFound(`generic resource "doesnotexist" doesn't exist`))
```

Comment:
Assert only that it is a NotFound? We don't need to go as far as an exact match of the error message, it is needlessly brittle.

Reply:
If this is not a right context to use `require.ErrorIs`, what is?

Reply:
My rule of thumb is checking the error message only for cases where there is ambiguity in the error type. In this case NotFound can only mean one thing, so checking the type is not ambiguous - on the contrary, it has the same practical effect and doesn't depend on any particular string.